### ISSUE TITLE
"Integer" fields return floats

### ISSUE THREAD

**Issue by organisciak** [Bookworm-project/BookwormAPI#14]
When I use "method: returnPossibleFields", it tells me that date_year is type: integer. However:

[{"date_year":{"$gte":0}}](http://bookworm.htrc.illinois.edu/cgi-bin/dbbindings.py?queryTerms={%22counttype%22:%20[%22TextCount%22],%20%22search_limits%22:%20[{%22date_year%22:%20{%22$gte%22:%200}}],%20%22groups%22:%20[%22date_year%22],%20%22database%22:%20%22hathipd%22,%20%22words_collation%22:%20%22Case_Sensitive%22,%20%22method%22:%20%22return_json%22,%20%22compare_limits%22:%20[]})

- All the date_year values are integers.

[{"date_year":{"$lte":2015}}](http://bookworm.htrc.illinois.edu/cgi-bin/dbbindings.py?queryTerms={%22counttype%22:%20[%22TextCount%22],%20%22search_limits%22:%20[{%22date_year%22:%20{%22$lte%22:%202015}}],%20%22groups%22:%20[%22date_year%22],%20%22database%22:%20%22hathipd%22,%20%22words_collation%22:%20%22Case_Sensitive%22,%20%22method%22:%20%22return_json%22,%20%22compare_limits%22:%20[]})

- All the date_year values have ".0". This causes python to yell at me when wrapped in int(): an easy fix for me, but presumably not something we should be getting back from the API.

**Comment by bmschmidt** [comment]
OK, can't fix this right away, but a couple notes:

Minimal example may be that 
[`"date_year":{"$gte":1}`](http://bookworm.htrc.illinois.edu/cgi-bin/dbbindings.py?queryTerms=%7B%22counttype%22:%20%5B%22TextCount%22%5D,%20%22search_limits%22:%20%5B%7B%22date_year%22:%20%7B%22%24gte%22:%200%7D%7D%5D,%20%22groups%22:%20%5B%22date_year%22%5D,%20%22database%22:%20%22hathipd%22,%20%22words_collation%22:%20%22Case_Sensitive%22,%20%22method%22:%20%22return_json%22,%20%22compare_limits%22:%20%5B%5D%7D)

The code that's supposed to be correcting this is [here](https://github.com/Bookworm-project/BookwormAPI/blob/master/bookworm/general_API.py#L347-L352): quoted,

``` {python}
        def fixNumpyType(input):
            #This is, weirdly, an occasional problem but not a constant one.
            if str(input.dtype)=="int64":
                return int(input)
            else:
                return input
```

I don't understand the internal types that numpy uses, but I suspect the problem is that we need to be coercing not just `int64`, but some of numpy's 11(!) other integer types to a standard int class before letting the json parser have at them. Maybe when the first number is zero, numpy gets different ideas what to do about a field.

**Comment by organisciak** [comment]
`issubclass(input.dtype.type, np.integer)` seems to work.

```python
import numpy as np

def printtype(a):
    print a.dtype
    print "\tINT?", issubclass(a.dtype.type, np.integer)
    print "\tFLOAT?", issubclass(a.dtype.type, np.float)

printtype(np.array([1,2,3]))
printtype(np.array([1,2,3], dtype="uint64"))
printtype(np.array([1.0,2,3]))
```
Output:
```
int32
        INT? True
        FLOAT? False
uint64
        INT? True
        FLOAT? False
float64
        INT? False
        FLOAT? True
```

**Comment by organisciak** [comment]
I added a test folder on bookworm.htrc.illinois.edu (simply add `test/` after `cgi-bin/`) and tried it out. It looks like fixNumpyType isn't even being called, though.

**Comment by bmschmidt** [comment]
Yeah, looks like I was wrong. When all numbers are coerced to int with the following code (which is unacceptable, because WordsPerMillion is a float), it does return ints. (Adding in the call to a few other places). So apparently either `numpy` or `mysqldb` is being a little finicky about when to return floats, and when ints. (Or, I suppose, it's even possible that some of these are **strings;** that seems pretty unlikely, but not much more than floats).

```python
        def fixNumpyType(input):
            #This is, weirdly, an occasional problem but not a constant one.                                                                                                      
            if issubclass(input.dtype.type, np.integer):
                return int(input)
            else:
                return int(input)
```
